# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: editor/libeditor/EditorBase.cpp
# Commit: 1466d7d61102
# Full Hash: 1466d7d6110279cf167ab8b1567be693df5d7c3c
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2020-03-03 09:50:30
# Regressor Bug: 1618089
# File Overlap Count: 2
# Description:
#   Bug 1618089 - part 7: Make some related methods use `EditorDOMPointInText` for arguments or result r=m_kato
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D64338
# ==============================================================================

diff -r 079f4b05ead2 -r 1466d7d61102 editor/libeditor/EditorBase.cpp
--- a/editor/libeditor/EditorBase.cpp	Tue Mar 03 02:19:59 2020 +0000
+++ b/editor/libeditor/EditorBase.cpp	Tue Mar 03 01:15:38 2020 +0000
@@ -36,11 +36,12 @@
 #include "mozilla/FlushType.h"                // for FlushType::Frames
 #include "mozilla/IMEContentObserver.h"       // for IMEContentObserver
 #include "mozilla/IMEStateManager.h"          // for IMEStateManager
-#include "mozilla/mozalloc.h"                 // for operator new, etc.
-#include "mozilla/mozInlineSpellChecker.h"    // for mozInlineSpellChecker
-#include "mozilla/mozSpellChecker.h"          // for mozSpellChecker
-#include "mozilla/Preferences.h"              // for Preferences
-#include "mozilla/PresShell.h"                // for PresShell
+#include "mozilla/InternalMutationEvent.h"  // for NS_EVENT_BITS_MUTATION_CHARACTERDATAMODIFIED
+#include "mozilla/mozalloc.h"               // for operator new, etc.
+#include "mozilla/mozInlineSpellChecker.h"  // for mozInlineSpellChecker
+#include "mozilla/mozSpellChecker.h"        // for mozSpellChecker
+#include "mozilla/Preferences.h"            // for Preferences
+#include "mozilla/PresShell.h"              // for PresShell
 #include "mozilla/RangeBoundary.h"      // for RawRangeBoundary, RangeBoundary
 #include "mozilla/Services.h"           // for GetObserverService
 #include "mozilla/ServoCSSParser.h"     // for ServoCSSParser
@@ -51,12 +52,12 @@
 #include "mozilla/TextServicesDocument.h"  // for TextServicesDocument
 #include "mozilla/TextEvents.h"
 #include "mozilla/TransactionManager.h"  // for TransactionManager
-#include "mozilla/dom/AbstractRange.h"   // for AbstractRange
-#include "mozilla/dom/CharacterData.h"   // for CharacterData
-#include "mozilla/dom/DataTransfer.h"    // for DataTransfer
-#include "mozilla/InternalMutationEvent.h"  // for NS_EVENT_BITS_MUTATION_CHARACTERDATAMODIFIED
-#include "mozilla/dom/Element.h"      // for Element, nsINode::AsElement
-#include "mozilla/dom/EventTarget.h"  // for EventTarget
+#include "mozilla/Tuple.h"
+#include "mozilla/dom/AbstractRange.h"  // for AbstractRange
+#include "mozilla/dom/CharacterData.h"  // for CharacterData
+#include "mozilla/dom/DataTransfer.h"   // for DataTransfer
+#include "mozilla/dom/Element.h"        // for Element, nsINode::AsElement
+#include "mozilla/dom/EventTarget.h"    // for EventTarget
 #include "mozilla/dom/HTMLBodyElement.h"
 #include "mozilla/dom/HTMLBRElement.h"
 #include "mozilla/dom/Selection.h"  // for Selection, etc.
@@ -2704,8 +2705,8 @@
       NS_ENSURE_TRUE(newOffset.isValid(), NS_ERROR_FAILURE);
     }
     nsresult rv = InsertTextIntoTextNodeWithTransaction(
-        aStringToInsert, MOZ_KnownLive(*pointToInsert.GetContainerAsText()),
-        pointToInsert.Offset());
+        aStringToInsert, EditorDOMPointInText(pointToInsert.ContainerAsText(),
+                                              pointToInsert.Offset()));
     NS_ENSURE_SUCCESS(rv, rv);
     if (aPointAfterInsertedString) {
       aPointAfterInsertedString->Set(pointToInsert.GetContainer(),
@@ -2719,8 +2720,8 @@
     NS_ENSURE_TRUE(newOffset.isValid(), NS_ERROR_FAILURE);
     // we are inserting text into an existing text node.
     nsresult rv = InsertTextIntoTextNodeWithTransaction(
-        aStringToInsert, MOZ_KnownLive(*pointToInsert.GetContainerAsText()),
-        pointToInsert.Offset());
+        aStringToInsert, EditorDOMPointInText(pointToInsert.ContainerAsText(),
+                                              pointToInsert.Offset()));
     NS_ENSURE_SUCCESS(rv, rv);
     if (aPointAfterInsertedString) {
       aPointAfterInsertedString->Set(pointToInsert.GetContainer(),
@@ -2762,56 +2763,48 @@
   return aString.EqualsLatin1(aTextFragment.Get1b() + aOffset, stringLength);
 }
 
-namespace {
-struct AdjustedInsertionRange {
-  EditorRawDOMPoint mBegin;
-  EditorRawDOMPoint mEnd;
-};
-}  // anonymous namespace
-
-static AdjustedInsertionRange AdjustTextInsertionRange(
-    Text& aTextNode, const int32_t aInsertionOffset,
-    const nsAString& aInsertedString) {
-  if (TextFragmentBeginsWithStringAtOffset(aTextNode.TextFragment(),
-                                           aInsertionOffset, aInsertedString)) {
-    EditorRawDOMPoint begin{&aTextNode, aInsertionOffset};
-    EditorRawDOMPoint end{
-        &aTextNode,
-        static_cast<int32_t>(aInsertionOffset + aInsertedString.Length())};
-    return {begin, end};
-  }
-
-  const EditorRawDOMPoint begin{&aTextNode, 0};
-  const EditorRawDOMPoint end{&aTextNode,
-                              static_cast<int32_t>(aTextNode.TextLength())};
-  return {begin, end};
+static Tuple<EditorDOMPointInText, EditorDOMPointInText>
+AdjustTextInsertionRange(const EditorDOMPointInText& aInsertedPoint,
+                         const nsAString& aInsertedString) {
+  if (TextFragmentBeginsWithStringAtOffset(
+          aInsertedPoint.ContainerAsText()->TextFragment(),
+          aInsertedPoint.Offset(), aInsertedString)) {
+    return MakeTuple(aInsertedPoint,
+                     EditorDOMPointInText(
+                         aInsertedPoint.ContainerAsText(),
+                         aInsertedPoint.Offset() + aInsertedString.Length()));
+  }
+
+  return MakeTuple(
+      EditorDOMPointInText(aInsertedPoint.ContainerAsText(), 0),
+      EditorDOMPointInText::AtEndOf(*aInsertedPoint.ContainerAsText()));
 }
 nsresult EditorBase::InsertTextIntoTextNodeWithTransaction(
-    const nsAString& aStringToInsert, Text& aTextNode, int32_t aOffset,
-    bool aSuppressIME) {
+    const nsAString& aStringToInsert,
+    const EditorDOMPointInText& aPointToInsert, bool aSuppressIME) {
   MOZ_ASSERT(IsEditActionDataAvailable());
-
+  MOZ_ASSERT(aPointToInsert.IsSetAndValid());
+
+  EditorDOMPointInText pointToInsert(aPointToInsert);
   RefPtr<EditTransactionBase> transaction;
   bool isIMETransaction = false;
-  RefPtr<Text> insertedTextNode = &aTextNode;
-  int32_t insertedOffset = aOffset;
   // aSuppressIME is used when editor must insert text, yet this text is not
   // part of the current IME operation. Example: adjusting whitespace around an
   // IME insertion.
   if (ShouldHandleIMEComposition() && !aSuppressIME) {
-    transaction = CompositionTransaction::Create(*this, aStringToInsert,
-                                                 aTextNode, aOffset);
+    transaction =
+        CompositionTransaction::Create(*this, aStringToInsert, pointToInsert);
     isIMETransaction = true;
     // All characters of the composition string will be replaced with
     // aStringToInsert.  So, we need to emulate to remove the composition
     // string.
     // FYI: The text node information in mComposition has been updated by
     //      CompositionTransaction::Create().
-    insertedTextNode = mComposition->GetContainerTextNode();
-    insertedOffset = mComposition->XPOffsetInTextNode();
+    pointToInsert.Set(mComposition->GetContainerTextNode(),
+                      mComposition->XPOffsetInTextNode());
   } else {
-    transaction = InsertTextTransaction::Create(*this, aStringToInsert,
-                                                aTextNode, aOffset);
+    transaction =
+        InsertTextTransaction::Create(*this, aStringToInsert, pointToInsert);
   }
 
   // XXX We may not need these view batches anymore.  This is handled at a
@@ -2820,22 +2813,22 @@
   nsresult rv = DoTransactionInternal(transaction);
   EndUpdateViewBatch();
 
-  if (AsHTMLEditor() && insertedTextNode) {
+  if (AsHTMLEditor() && pointToInsert.IsSet()) {
     // The DOM was potentially modified during the transaction. This is possible
     // through mutation event listeners. That is, the node could've been removed
     // from the doc or otherwise modified.
     if (!MaybeHasMutationEventListeners(
             NS_EVENT_BITS_MUTATION_CHARACTERDATAMODIFIED)) {
-      const EditorRawDOMPoint begin{insertedTextNode, insertedOffset};
-      const EditorRawDOMPoint end{
-          insertedTextNode,
-          static_cast<int32_t>(insertedOffset + aStringToInsert.Length())};
+      EditorDOMPointInText endOfInsertion(
+          pointToInsert.ContainerAsText(),
+          pointToInsert.Offset() + aStringToInsert.Length());
+      TopLevelEditSubActionDataRef().DidInsertText(*this, pointToInsert,
+                                                   endOfInsertion);
+    } else if (pointToInsert.ContainerAsText()->IsInComposedDoc()) {
+      EditorDOMPointInText begin, end;
+      Tie(begin, end) =
+          AdjustTextInsertionRange(pointToInsert, aStringToInsert);
       TopLevelEditSubActionDataRef().DidInsertText(*this, begin, end);
-    } else if (insertedTextNode->IsInComposedDoc()) {
-      AdjustedInsertionRange adjustedRange = AdjustTextInsertionRange(
-          *insertedTextNode, insertedOffset, aStringToInsert);
-      TopLevelEditSubActionDataRef().DidInsertText(*this, adjustedRange.mBegin,
-                                                   adjustedRange.mEnd);
     }
   }
 
@@ -2845,8 +2838,8 @@
     for (auto& listener : listeners) {
       // TODO: might need adaptation because of mutation event listeners called
       // during `DoTransactionInternal`.
-      listener->DidInsertText(insertedTextNode, insertedOffset, aStringToInsert,
-                              rv);
+      listener->DidInsertText(pointToInsert.ContainerAsText(),
+                              pointToInsert.Offset(), aStringToInsert, rv);
     }
   }
 